node/silkworm/trie/intermediate_hashes.cpp

```diff
@@ -18,133 +18,204 @@
 
 #include <bitset>
 
+#include <silkworm/common/endian.hpp>
 #include <silkworm/common/log.hpp>
 #include <silkworm/common/rlp_err.hpp>
 #include <silkworm/db/tables.hpp>
 
 namespace silkworm::trie {
 
-AccountTrieCursor::AccountTrieCursor(mdbx::txn& txn, PrefixSet& changed)
-    : changed_{changed}, cursor_{db::open_cursor(txn, db::table::kTrieOfAccounts)} {}
+Cursor::Cursor(mdbx::cursor& cursor, PrefixSet& changed, ByteView prefix)
+    : cursor_{cursor}, changed_{changed}, prefix_{prefix} {
+    consume_node(/*key=*/{}, /*exact=*/true);
+}
+
+void Cursor::consume_node(ByteView to, bool exact) {
+    const Bytes db_key{prefix_ + Bytes{to}};
+    const auto entry{exact ? cursor_.find(db::to_slice(db_key), /*throw_notfound=*/false)
+                           : cursor_.lower_bound(db::to_slice(db_key), /*throw_notfound=*/false)};
 
-void AccountTrieCursor::consume_node(ByteView to) {
-    const auto entry{cursor_.lower_bound(db::to_slice(to), /*throw_notfound=*/false)};
-    if (!entry) {
+    if (!entry && !exact) {
         // end-of-tree
-        root_nibble_ = 16;
+        stack_.clear();
         return;
     }
 
-    const auto node{unmarshal_node(db::from_slice(entry.value))};
-    assert(node != std::nullopt);
-    assert(node->state_mask() != 0);
-    uint8_t nibble{0};
-    while ((node->state_mask() & (1u << nibble)) == 0) {
-        ++nibble;
+    ByteView key = to;
+    if (!exact) {
+        key = db::from_slice(entry.key);
+        if (!has_prefix(key, prefix_)) {
+            stack_.clear();
+            return;
+        }
+        key.remove_prefix(prefix_.length());
+    }
+
+    std::optional<Node> node{std::nullopt};
+    if (entry) {
+        node = unmarshal_node(db::from_slice(entry.value));
+        assert(node != std::nullopt);
+        assert(node->state_mask() != 0);
```

Comment:
Not sure about this.
Is it expected that when entry then node MUST be valued ? If so the assertion is not enough IMHO as it gets voided in release build.

Reply:
Fair enough. https://github.com/torquem-ch/silkworm/pull/488/commits/c4a710cfc8ae1535b854a939b3d5c2b43373a2b1 introduces  `SILKWORM_ASSERT` that always aborts program execution on assertion failure.